mono/metadata/w32handle.c

```diff
@@ -939,29 +947,46 @@ mono_w32handle_wait_multiple (gpointer *handles, gsize nhandles, gboolean waital
 		{
 			mono_trace (G_LOG_LEVEL_DEBUG, MONO_TRACE_IO_LAYER_HANDLE, "%s: handle %p can't be waited for", __func__, handles_data [i]);
 
-			for (i = nhandles - 1; i >= 0; --i)
-				mono_w32handle_unref (handles_data [i]);
-
-			return MONO_W32HANDLE_WAIT_RET_FAILED;
+			ret = MONO_W32HANDLE_WAIT_RET_FAILED;
+			goto done;
 		}
 
 		handles_data_sorted [i] = handles_data [i];
 	}
 
+	// Duplication is ok for WaitAny, exception for WaitAll.
+	// System.DuplicateWaitObjectException: Duplicate objects in argument.
+	// It is not obviously sensible to otherwise sort/unique the handles,
+	// as we have to return the lowest signaled one when multiple are signaled.
+	// Lowest in terms of array index in the input array.
 	qsort (handles_data_sorted, nhandles, sizeof (gpointer), g_direct_equal);
 	for (i = 1; i < nhandles; ++i) {
 		if (handles_data_sorted [i - 1] == handles_data_sorted [i]) {
-			mono_trace (G_LOG_LEVEL_DEBUG, MONO_TRACE_IO_LAYER_HANDLE, "%s: handle %p is duplicated", __func__, handles_data_sorted [i]);
-
-			for (i = nhandles - 1; i >= 0; --i)
-				mono_w32handle_unref (handles_data [i]);
-
-			return MONO_W32HANDLE_WAIT_RET_FAILED;
+			if (waitall) {
+				mono_error_set_duplicate_wait_object (error);
+				mono_trace (G_LOG_LEVEL_DEBUG, MONO_TRACE_IO_LAYER_HANDLE, "%s: handle %p is duplicated", __func__, handles_data_sorted [i]);
+				ret = MONO_W32HANDLE_WAIT_RET_FAILED;
+				goto done;
+			} else {
+				// There is are duplicates. Remove them -- in-place in
+				// order to return the lowest signaled, equal to the caller's
+				// indices, and ease the exit path's dereference.
+				for (i = 0; i < nhandles; ++i) {
+					for (gsize j = i + 1; j < nhandles; ++j) {
+						if (handles_data [i] == handles_data [j]) {
+							mono_w32handle_unref (handles_data [j]);
+							handles_data [j] = NULL;
+						}
```

Comment:
Actually that data isn't sorted, and operating on the sorted data is tricky, so the code is a bit inefficient but about what it should be. Duplicates will be rare. Worst case is 64x64 but usually much less. The initial check for duplicates will tend to make this entirely skipped, esp. since mono didn't handle this before.

Reply:
You have 2 choices here to avoid the inner-loop:
 - if growing upward (from i = 1 to nhandles), you can just set the `handles_data_sorted [i - 1]` to `NULL` so if `handles_data_sorted [i + 1]` is also a duplicate you haven't changed `handles_data_sorted [i]`
 - if growing downward (from i = nhandles - 1 to 0), you can still set `handles_data_sorted [i]` to `NULL` when comparing it to `handles_data_sorted [i + 1]` so when looking at `handles_data_sorted [i - 1]`, `handles_data_sorted [i]` haven't been changed.

I would go for the first one since it's the minimal change and the result is the same.

Reply:
I missed the fact that you are checking `handles_data_sorted` but modifying `handles_data`. Extracting the duplication removal would certainly help in making the code clearer. In simpler terms: please move https://github.com/jaykrell/mono/blob/4493213cd1cc722603e838eb1e5b9977d404ff91/mono/metadata/w32handle.c?utf8=✓#L964-L988 after https://github.com/mono/mono/pull/9114/files#diff-81316870ec3e337f2fd9f434dfbd0a4cR962 loop, using a `duplicated` variable for example.

Reply:
There is really no need to move it. We can do the "action" right after the "detection". There is only one instance of each so no need to flow multiple  "detections" into the one "action". If a variable is introduced, the human reader has to go searching for all the reads and writes, only to find one of each and wonder why the code is split apart. The comments explain everything pretty well here I think.

Reply:
Well, I didn't. What makes sense to you might not makes sense to other people and vice-versa.